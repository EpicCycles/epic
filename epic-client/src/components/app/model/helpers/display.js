export const fixedHeaderClassname = (lockColumn) => {
    if (lockColumn) return "grid-header--fixed-left";
    return "";
};
export const fixedDetailsClassname = (lockColumn) => {
    if (lockColumn) return "grid-item--fixed-left";
    return "";
};
export const gridHeaderClass = (baseClassName = "", fieldIndex, firstColumnLocked) => {
  const shouldLock = firstColumnLocked && (fieldIndex === 0);
    return `${baseClassName} grid-item--header ${fixedHeaderClassname(shouldLock)}`;
};
export const gridItemClass = (baseClassName = "", fieldIndex, firstColumnLocked) => {
  const shouldLock = firstColumnLocked && (fieldIndex === 0);
    return `${baseClassName} grid-item ${fixedDetailsClassname(shouldLock)}`;
};