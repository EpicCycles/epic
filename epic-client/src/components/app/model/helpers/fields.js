import {
    ADDRESS_MISSING,
    BRAND_MISSING, BRAND_NAME_MISSING,
    BUNDLE_NAME_MISSING, COUNTRY_MISSING,
    FRAME_NAME_MISSING,
    MODEL_NAME_MISSING,
    PART_MISSING,
    PART_NAME_MISSING,
    PART_TYPE_MISSING,
    PRODUCTS_MISSING,
    SUPPLIER_MISSING
} from "./error";
import {validatePostcodeAndReturnError} from "./validators";

// TODO customer fields including email with email validator
export const ADDRESS1 = "address1";
export const ADDRESS2 = "address2";
export const ADDRESS3 = "address3";
export const ADDRESS4 = "address4";
export const COUNTRY = "country";
export const POSTCODE = "postcode";
export const BRAND = "brand";
export const SUPPLIER = "supplier";
export const PART = "part";
export const LINK = "link";
export const BUNDLE = "bundle_name";
export const PRODUCT_CODE = "product_code";
export const PART_TYPE = "partType";
export const BRAND_NAME = "brand_name";
export const BIKE_BRAND = "bike_brand";
export const FRAME_NAME = "frame_name";
export const MODEL_NAME = "model_name";
export const PART_NAME = "part_name";
export const DESCRIPTION = "description";
export const COLOURS = "colours";
export const SELL_PRICE = "rrp";
export const EPIC_PRICE = "epic_price";
export const TICKET_PRICE = "ticket_price";
export const TRADE_PRICE = "trade_price";
export const CLUB_PRICE = "club_price";
export const FITTED_PRICE = "fitted_price";
export const TRADE_IN_PRICE = "trade_in_value";
export const SIZES = "sizes";
export const STANDARD = "standard";
export const STOCKED = "stocked";
export const PRODUCTS = "products";
export const TEXT = "text";
export const NUMBER = "number";
export const CURRENCY = "currency";
export const TEXT_AREA = "textArea";
export const CHECKBOX = "checkbox";

// TODO add COUNTRY type and fields for address
// TODO consider having a function passed for validator
export const ADDRESS1_FIELD = {
    fieldName: ADDRESS1,
    type: TEXT,
    length: 100,
    header: "Address",
    required: true,
    error: ADDRESS_MISSING
};
export const ADDRESS2_FIELD = {
    fieldName: ADDRESS2,
    type: TEXT,
    length: 100,
    header: "Line 2",
};
export const ADDRESS3_FIELD = {
    fieldName: ADDRESS3,
    type: TEXT,
    length: 100,
    header: "Line 3",
};
export const ADDRESS4_FIELD = {
    fieldName: ADDRESS4,
    type: TEXT,
    length: 100,
    header: "Line 4",
};
export const COUNTRY_FIELD = {
    fieldName: COUNTRY,
    header: "Country",
    type: COUNTRY,
    required: true,
    error: COUNTRY_MISSING
};
export const POSTCODE_FIELD = {
    fieldName: POSTCODE,
    type: TEXT,
    length: 20,
    header: "Postcode",
    validator: validatePostcodeAndReturnError,
    validatorAdditionalFields: [COUNTRY]
};
export const BRAND_FIELD = {
    fieldName: BRAND,
    header: "Brand",
    synonyms: [BRAND, 'manufacturer'],
    required: true,
    error: BRAND_MISSING,
    type: BRAND
};
export const PRODUCTS_FIELD = {
    fieldName: PRODUCTS,
    header: "Products",
    synonyms: [],
    required: true,
    error: PRODUCTS_MISSING,
    type: PRODUCTS
};
export const SUPPLIER_FIELD = {
    fieldName: SUPPLIER,
    header: "Supplier",
    synonyms: [SUPPLIER],
    required: true,
    error: SUPPLIER_MISSING,
    type: SUPPLIER
};
export const SUPPLIER_FIELD_OPTIONAL = {
    fieldName: SUPPLIER,
    header: "Supplier",
    synonyms: [SUPPLIER],
    type: SUPPLIER
};
export const PART_FIELD = {
    fieldName: PART,
    header: "Part",
    synonyms: [PART],
    required: true,
    error: PART_MISSING,
    type: PART
};
export const FRAME_NAME_FIELD = {
    fieldName: FRAME_NAME,
    header: "Frame Name",
    synonyms: [],
    required: true,
    error: FRAME_NAME_MISSING
};
export const LINK_FIELD = {
    fieldName: LINK,
    header: "URL",
    synonyms: [],
};
export const BRAND_NAME_FIELD = {
    fieldName: BRAND_NAME,
    header: "Brand Name",
    synonyms: [],
    required: true,
    error: BRAND_NAME_MISSING
};
export const BUNDLE_NAME_FIELD = {
    fieldName: BUNDLE,
    header: "Bundle Name",
    synonyms: [],
    required: true,
    error: BUNDLE_NAME_MISSING
};
export const MODEL_NAME_FIELD = {
    fieldName: MODEL_NAME,
    header: "Model Name",
    synonyms: [],
    required: true,
    error: MODEL_NAME_MISSING,
    type: TEXT,
    length: 100
};
export const COLOURS_FIELD = {
    fieldName: COLOURS,
    header: "Colours",
    synonyms: [COLOURS, "colour", "colors", "color"],
    type: TEXT,
    length: 100
};
export const DESCRIPTION_FIELD = {
    fieldName: DESCRIPTION,
    header: "Description",
    synonyms: [DESCRIPTION, "desc"],
    type: TEXT_AREA,
    length: 400
};
export const PRODUCT_CODE_FIELD = {
    fieldName: PRODUCT_CODE,
    header: "Product Code",
    synonyms: [PRODUCT_CODE],
    type: TEXT,
    length: 30
};
export const SELL_PRICE_FIELD = {
    fieldName: SELL_PRICE,
    header: "RRP",
    synonyms: ["price", "selling price", "srp", "rrp", "sell price", "retail price"],
    type: CURRENCY,
    length: 10
};
export const EPIC_PRICE_FIELD = {
    fieldName: EPIC_PRICE,
    header: "Epic Price",
    synonyms: [],
    type: CURRENCY,
    length: 10
};
export const CLUB_PRICE_FIELD = {
    fieldName: CLUB_PRICE,
    header: "Club Price",
    synonyms: [],
    type: CURRENCY,
    length: 10
};

export const FITTED_PRICE_FIELD = {
    fieldName: FITTED_PRICE,
    header: "Fitted Price",
    synonyms: [],
    type: CURRENCY,
    length: 10
};
export const TRADE_PRICE_FIELD = {
    fieldName: TRADE_PRICE,
    header: "Trade Price",
    synonyms: [],
    type: CURRENCY,
    length: 10
};
export const TICKET_PRICE_FIELD = {
    fieldName: TICKET_PRICE,
    header: "Ticket Price",
    synonyms: [],
    type: CURRENCY,
    length: 10
};
export const SIZES_FIELD = {
    fieldName: SIZES,
    header: "Sizes",
    synonyms: [SIZES, "size", "frame sizes", "frame size"],
    type: TEXT,
    length: 100
};
export const PART_TYPE_FIELD = {
    fieldName: PART_TYPE,
    header: "Part Type",
    required: true,
    error: PART_TYPE_MISSING,
    type: PART_TYPE,
};
export const PART_NAME_FIELD = {
    fieldName: PART_NAME,
    header: "Part Name",
    required: true,
    error: PART_NAME_MISSING
};
export const TRADE_IN_FIELD = {
    fieldName: TRADE_IN_PRICE,
    header: "Trade In Â£",
    type: CURRENCY,
    length: 10
};
export const STOCKED_FIELD = {
    fieldName: STOCKED,
    header: "Stocked",
    type: CHECKBOX
};
export const STANDARD_FIELD = {
    fieldName: STANDARD,
    header: "Standard",
    type: CHECKBOX
};
export const BIKE_BRAND_FIELD = {
    fieldName: BIKE_BRAND,
    header: "Bike Brand",
    type: CHECKBOX
};
export const customerAddressFields = [ADDRESS1_FIELD, ADDRESS2_FIELD, ADDRESS3_FIELD, ADDRESS4_FIELD, COUNTRY_FIELD, POSTCODE_FIELD];
export const frameFields = [BRAND, FRAME_NAME];
export const brandFields = [BRAND_NAME_FIELD, BIKE_BRAND_FIELD, LINK_FIELD, SUPPLIER_FIELD_OPTIONAL];
export const bikeFields = [MODEL_NAME_FIELD, DESCRIPTION_FIELD, COLOURS_FIELD, SELL_PRICE_FIELD, EPIC_PRICE_FIELD, CLUB_PRICE_FIELD, SIZES_FIELD];
export const partFields = [
    PART_TYPE_FIELD,
    BRAND_FIELD,
    PART_NAME_FIELD,
    TRADE_IN_FIELD,
    STANDARD_FIELD,
    STOCKED_FIELD
];
export const supplierProductFields = [
    SUPPLIER_FIELD,
    FITTED_PRICE_FIELD,
    TICKET_PRICE_FIELD,
    SELL_PRICE_FIELD,
    TRADE_PRICE_FIELD,
    CLUB_PRICE_FIELD
];
export const bundleFields = [
    BUNDLE_NAME_FIELD,
    PRODUCTS_FIELD,
    FITTED_PRICE_FIELD,
    TICKET_PRICE_FIELD,
];

