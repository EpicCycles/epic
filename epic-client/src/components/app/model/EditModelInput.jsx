import * as PropTypes from "prop-types";
import React, {Component, Fragment} from "react";
import {BRAND, CHECKBOX, COUNTRY, CURRENCY, NUMBER, PART_TYPE, SUPPLIER, TEXT_AREA} from "../../../helpers/models";
import FormTextAreaInput from "../../../common/FormTextAreaInput";
import FormTextInput from "../../../common/FormTextInput";
import PartTypeSelect from "../../partType/PartTypeSelect";
import BrandSelect from "../../brand/BrandSelect";
import SupplierSelect from "../../supplier/SupplierSelect";
import CountrySelect from "../../../common/CountrySelect";

class EditModelInput extends Component {
    validateOnChange = (fieldName, fieldValue) => {
        if (fieldValue) {
            this.props.onChange(this.props.field.fieldName, fieldValue, this.props.componentKey);
        } else {
            this.props.onChange(this.props.field.fieldName, "", this.props.componentKey);
        }
    };
    resetField = fieldName => {
        const originalValue = this.props.persistedModel ? this.props.persistedModel[this.props.field.fieldName] : undefined;
        this.validateOnChange(fieldName, originalValue);
    };

    render() {
        const { field, model, className, componentKey, index, sections, brands, suppliers } = this.props;
        let editComponent;
        const fieldName = `${field.fieldName}_${componentKey}${index}`;
        const fieldValue = model && model[field.fieldName];
        const emptyAllowed = !(field.required && fieldValue);
        const error = model.error_detail ? model.error_detail[field.fieldName] : "";

        // TODO add country type field
        switch (field.type) {
            case TEXT_AREA:
                editComponent = <FormTextAreaInput
                    className={className}
                    placeholder={field.header}
                    fieldName={fieldName}
                    value={fieldValue}
                    onChange={this.validateOnChange}
                    cols={Math.ceil(field.size / 4)}
                    onClick={this.resetField}
                    error={error}
                />;
                break;
            case COUNTRY:
                editComponent = <CountrySelect
                    fieldName={fieldName}
                    countrySelected={fieldValue}
                    onChange={this.validateOnChange}
                    isEmptyAllowed={emptyAllowed}
                />;
                break;
            case NUMBER:
            case CURRENCY:
                editComponent = <FormTextInput
                    type="number"
                    className={className}
                    placeholder={field.header}
                    error={error}
                    fieldName={fieldName}
                    value={fieldValue}
                    onChange={this.validateOnChange}
                    size={field.size}
                    onClick={this.validateOnChange}
                />;
                break;
            case CHECKBOX:
                editComponent = <div className="ui toggle checkbox">
                    <input type="checkbox"
                           name={fieldName}
                           onChange={() => this.validateOnChange(fieldName, !fieldValue)}
                           checked={fieldValue}
                           className={error ? "red" : ""}
                           title={error ? `${field.title} red` : field.title}
                    /></div>;
                break;
            case PART_TYPE:
                editComponent = <PartTypeSelect
                    sections={sections}
                    fieldName={fieldName}
                    onChange={this.validateOnChange}
                    partTypeSelected={fieldValue}
                    isEmptyAllowed={emptyAllowed}
                    error={error}
                    disabled={false}
                />;
                break;
            case BRAND:
                editComponent = <BrandSelect
                    brands={brands}
                    fieldName={fieldName}
                    onChange={this.validateOnChange}
                    brandSelected={fieldValue}
                    isEmptyAllowed={emptyAllowed}
                    error={error}
                    disabled={false}
                    bikeOnly={field.bikeOnly}
                />;
                break;
            case SUPPLIER:
                editComponent = <SupplierSelect
                    suppliers={suppliers}
                    fieldName={fieldName}
                    onChange={this.validateOnChange}
                    supplierSelected={fieldValue}
                    isEmptyAllowed={emptyAllowed}
                    error={error}
                    disabled={false}
                />;
                break;
            default:
                editComponent = <FormTextInput
                    className={className}
                    placeholder={field.header}
                    error={error}
                    onChange={this.validateOnChange}
                    fieldName={fieldName}
                    value={fieldValue}
                    size={field.size}
                    onClick={this.validateOnChange}
                />;
        }
        return <Fragment>{editComponent}</Fragment>;
    };

}

EditModelInput.propTypes = {
    field: PropTypes.object.isRequired,
    model: PropTypes.object.isRequired,
    persistedModel: PropTypes.object,
    className: PropTypes.string,
    componentKey: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.number,
    ]).isRequired,
    index: PropTypes.number.isRequired,
    sections: PropTypes.array,
    brands: PropTypes.array,
    suppliers: PropTypes.array,
    onChange: PropTypes.func.isRequired
};
export default EditModelInput;