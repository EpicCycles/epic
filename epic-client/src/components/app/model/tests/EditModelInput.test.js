import React from "react";
import {BRAND, CHECKBOX, COUNTRY, CURRENCY, PART_TYPE, SUPPLIER, TEXT, TEXT_AREA} from "../helpers/fields";
import EditModelInput from "../EditModelInput";

const foundName = "find me";
const sections = [
    {
        id: 1,
        partTypes: [
            { id: 11, name: "id 11" },
            { id: 21, name: "id 11" },
        ]
    },
    {
        id: 2,
        partTypes: [
            { id: 2, name: foundName },
            { id: 22, name: "id 11" },
        ]
    },
];
const brands = [
    { id: 1, brand_name: "id is 1" },
    { id: 2, brand_name: foundName },
    { id: 3, brand_name: "id is 3" },
];
const suppliers = [
    { id: 1, supplier_name: "id is 1" },
    { id: 2, supplier_name: foundName },
    { id: 3, supplier_name: "id is 3" },
];
const emptyModel = {};
const componentKey = 12;
test("it renders a currency field that has data", () => {
    const field = {
        fieldName: "data_field",
        type: CURRENCY,
        length: 10
    };
    const model = { data_field: 23.90 };
    expect(shallow(<EditModelInput
        field={field}
        model={model}
        componentKey={componentKey}
        index={0}
        onChange={jest.fn()}
    />)).toMatchSnapshot();
});
test("it renders a currency field that has no data", () => {
    const field = {
        fieldName: "data_field",
        type: CURRENCY,
        length: 10
    };
    expect(shallow(<EditModelInput
        field={field}
        model={emptyModel}
        componentKey={componentKey}
        index={0}
        onChange={jest.fn()}
    />)).toMatchSnapshot();
});
test("it renders a country field that has data", () => {
    const field = {
        fieldName: "data_field",
        type: COUNTRY,
    };
    const model = { data_field: "NZ" };
    expect(shallow(<EditModelInput
        field={field}
        model={model}
        componentKey={componentKey}
        index={0}
        onChange={jest.fn()}
    />)).toMatchSnapshot();
});
test("it renders a country field that has no data", () => {
    const field = {
        fieldName: "data_field",
        type: COUNTRY,
    };
    expect(shallow(<EditModelInput
        field={field}
        model={emptyModel}
        componentKey={componentKey}
        index={0}
        onChange={jest.fn()}
    />)).toMatchSnapshot();
});
test("it renders a text field that has data", () => {
    const field = {
        fieldName: "data_field",
        type: TEXT,
        length: 10
    };
    const model = { data_field: "SHow text" };
    expect(shallow(<EditModelInput
        field={field}
        model={model}
        componentKey={componentKey}
        index={0}
        onChange={jest.fn()}
    />)).toMatchSnapshot();
});
test("it renders a text field that has no data", () => {
    const field = {
        fieldName: "data_field",
        type: TEXT,
        length: 10
    };
    expect(shallow(<EditModelInput
        field={field}
        model={emptyModel}
        componentKey={componentKey}
        index={0}
        onChange={jest.fn()}
    />)).toMatchSnapshot();
});
test("it renders a textarea field that has data", () => {
    const field = {
        fieldName: "data_field",
        type: TEXT_AREA,
        size: 400,
    };
    const model = { data_field: "SHow text" };
    expect(shallow(<EditModelInput
        field={field}
        model={model}
        componentKey={componentKey}
        index={0}
        onChange={jest.fn()}
    />)).toMatchSnapshot();
});
test("it renders a textarea field that has no data", () => {
    const field = {
        fieldName: "data_field",
        type: TEXT_AREA,
        size: 400,
        header: "put text here"
    };
    expect(shallow(<EditModelInput
        field={field}
        model={emptyModel}
        componentKey={componentKey}
        index={0}
        onChange={jest.fn()}
    />)).toMatchSnapshot();
});

test("it renders a checkbox field that is true", () => {
    const field = {
        fieldName: "data_field",
        type: CHECKBOX,
    };
    const model = { data_field: true };
    expect(shallow(<EditModelInput
        field={field}
        model={model}
        componentKey={componentKey}
        index={0}
        onChange={jest.fn()}
    />)).toMatchSnapshot();
});
test("it renders a checkbox field that is false", () => {
    const field = {
        fieldName: "data_field",
        type: CHECKBOX,
    };
    const model = { data_field: false };
    expect(shallow(<EditModelInput
        field={field}
        model={model}
        componentKey={componentKey}
        index={0}
        onChange={jest.fn()}
    />)).toMatchSnapshot();
});
test("it renders a checkbox field that has no data", () => {
    const field = {
        fieldName: "data_field",
        type: CURRENCY,
        length: 10
    };
    expect(shallow(<EditModelInput
        field={field}
        model={emptyModel}
        componentKey={componentKey}
        index={0}
        onChange={jest.fn()}
    />)).toMatchSnapshot();
});
test("it renders a part type field that has data that is found", () => {
    const field = {
        fieldName: "data_field",
        type: PART_TYPE,
        length: 10
    };
    const model = { data_field: 2 };
    expect(shallow(<EditModelInput
        field={field}
        model={model}
        componentKey={componentKey}
        index={0}
        onChange={jest.fn()}
        sections={sections}
    />)).toMatchSnapshot();
});
test("it renders a part type field that has data that is not found", () => {
    const field = {
        fieldName: "data_field",
        type: PART_TYPE,
        length: 10
    };
    const model = { data_field: 202 };

    expect(shallow(<EditModelInput
        field={field}
        model={model}
        componentKey={componentKey}
        index={0}
        onChange={jest.fn()}
        sections={sections}
    />)).toMatchSnapshot();
});
test("it renders a part Type field that has no data", () => {
    const field = {
        fieldName: "data_field",
        type: PART_TYPE,
        length: 10
    };
    expect(shallow(<EditModelInput
        field={field}
        model={emptyModel}
        componentKey={componentKey}
        index={0}
        onChange={jest.fn()}
        sections={sections}
    />)).toMatchSnapshot();
});
test("it renders a brand field that has data that is found", () => {
    const field = {
        fieldName: "data_field",
        type: BRAND,
        length: 10
    };
    const model = { data_field: 2 };
    expect(shallow(<EditModelInput
        field={field}
        model={model}
        componentKey={componentKey}
        index={0}
        onChange={jest.fn()}
        brands={brands}
    />)).toMatchSnapshot();
});
test("it renders a brand field that has data that is not found", () => {
    const field = {
        fieldName: "data_field",
        type: BRAND,
    };
    const model = { data_field: 202 };

    expect(shallow(<EditModelInput
        field={field}
        model={model}
        componentKey={componentKey}
        index={0}
        onChange={jest.fn()}
        brands={brands}
    />)).toMatchSnapshot();
});
test("it renders a brand field that has no data", () => {
    const field = {
        fieldName: "data_field",
        type: BRAND,
    };
    expect(shallow(<EditModelInput
        field={field}
        model={emptyModel}
        componentKey={componentKey}
        index={0}
        onChange={jest.fn()}
        brands={brands}
    />)).toMatchSnapshot();
});
test("it renders a supplier field that has data that is found", () => {
    const field = {
        fieldName: "data_field",
        type: SUPPLIER,
        length: 10
    };
    const model = { data_field: 2 };
    expect(shallow(<EditModelInput
        field={field}
        model={model}
        componentKey={componentKey}
        index={0}
        onChange={jest.fn()}
        suppliers={suppliers}
    />)).toMatchSnapshot();
});
test("it renders a supplier field that has data that is not found", () => {
    const field = {
        fieldName: "data_field",
        type: SUPPLIER,
    };
    const model = { data_field: 202 };

    expect(shallow(<EditModelInput
        field={field}
        model={model}
        componentKey={componentKey}
        index={0}
        onChange={jest.fn()}
        suppliers={suppliers}
    />)).toMatchSnapshot();
});
test("it renders a supplier field that has no data", () => {
    const field = {
        fieldName: "data_field",
        type: SUPPLIER,
    };
    expect(shallow(<EditModelInput
        field={field}
        model={emptyModel}
        componentKey={componentKey}
        index={0}
        onChange={jest.fn()}
        suppliers={suppliers}
    />)).toMatchSnapshot();
});
test("it calls onchange with a valid field setting state correctly", () => {
    const field = {
        fieldName: "data_field",
        type: CURRENCY,
        length: 10
    };
    const model = { data_field: 23.90 };
    const persistedModel = { data_field: 123.99 };
    let onChange = jest.fn();
    let component = shallow(<EditModelInput
        field={field}
        model={model}
        persistedModel={persistedModel}
        componentKey={componentKey}
        index={0}
        onChange={onChange}
    />);
    component.instance().validateOnChange(field.fieldName, '45.75');
    expect(onChange.mock.calls.length).toBe(1);
    expect(onChange.mock.calls[0][0]).toBe(field.fieldName);
    expect(onChange.mock.calls[0][1]).toBe('45.75');
    expect(onChange.mock.calls[0][2]).toBe(componentKey);
});
test("it calls reset on a field to set back to model value", () => {
    const field = {
        fieldName: "data_field",
        type: CURRENCY,
        length: 10
    };
    const model = { data_field: 23.90 };
    const persistedModel = { data_field: 123.99 };
    let onChange = jest.fn();
    let component = shallow(<EditModelInput
        field={field}
        model={model}
        persistedModel={persistedModel}
        componentKey={componentKey}
        index={0}
        onChange={onChange}
    />);
    component.instance().resetField(field.fieldName);
    expect(onChange.mock.calls.length).toBe(1);
    expect(onChange.mock.calls[0][0]).toBe(field.fieldName);
    expect(onChange.mock.calls[0][1]).toBe(persistedModel.data_field);
    expect(onChange.mock.calls[0][2]).toBe(componentKey);
});
test("it calls reset on a field with no model and shows error", () => {
    const field = {
        fieldName: "data_field",
        type: CURRENCY,
        length: 10,
        required: true,
        error: "My data is missing"
    };
    const model = { data_field: 23.90 };
    let onChange = jest.fn();
    let component = shallow(<EditModelInput
        field={field}
        model={model}
        persistedModel={emptyModel}
        componentKey={componentKey}
        index={0}
        onChange={onChange}
    />);
    component.instance().resetField(field.fieldName);
    expect(onChange.mock.calls.length).toBe(1);
    expect(onChange.mock.calls[0][0]).toBe(field.fieldName);
    expect(onChange.mock.calls[0][1]).toBe("");
    expect(onChange.mock.calls[0][2]).toBe(componentKey);
});
test("it calls onchange with a missing field setting state correctly", () => {
    const field = {
        fieldName: "data_field",
        type: CURRENCY,
        length: 10,
        required: true,
        error: "My data is missing"
    };
    const model = { data_field: 23.90 };
    const persistedModel = { data_field: 123.99 };
    let onChange = jest.fn();
    let component = shallow(<EditModelInput
        field={field}
        model={model}
        persistedModel={persistedModel}
        componentKey={componentKey}
        index={0}
        onChange={onChange}
    />);
    component.instance().validateOnChange(field.fieldName);
    expect(onChange.mock.calls.length).toBe(1);
    expect(onChange.mock.calls[0][0]).toBe(field.fieldName);
    expect(onChange.mock.calls[0][1]).toBe("");
    expect(onChange.mock.calls[0][2]).toBe(componentKey);
});