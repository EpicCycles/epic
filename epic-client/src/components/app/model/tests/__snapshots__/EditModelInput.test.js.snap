// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders a brand field that has data that is found 1`] = `
<React.Fragment>
  <BrandSelect
    brandSelected={2}
    brands={
      Array [
        Object {
          "brand_name": "id is 1",
          "id": 1,
        },
        Object {
          "brand_name": "find me",
          "id": 2,
        },
        Object {
          "brand_name": "id is 3",
          "id": 3,
        },
      ]
    }
    disabled={false}
    error=""
    fieldName="data_field_120"
    isEmptyAllowed={true}
    onChange={[Function]}
  />
</React.Fragment>
`;

exports[`it renders a brand field that has data that is not found 1`] = `
<React.Fragment>
  <BrandSelect
    brandSelected={202}
    brands={
      Array [
        Object {
          "brand_name": "id is 1",
          "id": 1,
        },
        Object {
          "brand_name": "find me",
          "id": 2,
        },
        Object {
          "brand_name": "id is 3",
          "id": 3,
        },
      ]
    }
    disabled={false}
    error=""
    fieldName="data_field_120"
    isEmptyAllowed={true}
    onChange={[Function]}
  />
</React.Fragment>
`;

exports[`it renders a brand field that has no data 1`] = `
<React.Fragment>
  <BrandSelect
    brands={
      Array [
        Object {
          "brand_name": "id is 1",
          "id": 1,
        },
        Object {
          "brand_name": "find me",
          "id": 2,
        },
        Object {
          "brand_name": "id is 3",
          "id": 3,
        },
      ]
    }
    disabled={false}
    error=""
    fieldName="data_field_120"
    isEmptyAllowed={true}
    onChange={[Function]}
  />
</React.Fragment>
`;

exports[`it renders a checkbox field that has no data 1`] = `
<React.Fragment>
  <FormTextInput
    error=""
    fieldName="data_field_120"
    onChange={[Function]}
    onClick={[Function]}
    type="number"
  />
</React.Fragment>
`;

exports[`it renders a checkbox field that is false 1`] = `
<React.Fragment>
  <div
    className="ui toggle checkbox"
  >
    <input
      checked={false}
      className=""
      name="data_field_120"
      onChange={[Function]}
      type="checkbox"
    />
  </div>
</React.Fragment>
`;

exports[`it renders a checkbox field that is true 1`] = `
<React.Fragment>
  <div
    className="ui toggle checkbox"
  >
    <input
      checked={true}
      className=""
      name="data_field_120"
      onChange={[Function]}
      type="checkbox"
    />
  </div>
</React.Fragment>
`;

exports[`it renders a country field that has data 1`] = `
<React.Fragment>
  <CountrySelect
    countrySelected="NZ"
    fieldName="data_field_120"
    isEmptyAllowed={true}
    onChange={[Function]}
  />
</React.Fragment>
`;

exports[`it renders a country field that has no data 1`] = `
<React.Fragment>
  <CountrySelect
    fieldName="data_field_120"
    isEmptyAllowed={true}
    onChange={[Function]}
  />
</React.Fragment>
`;

exports[`it renders a currency field that has data 1`] = `
<React.Fragment>
  <FormTextInput
    error=""
    fieldName="data_field_120"
    onChange={[Function]}
    onClick={[Function]}
    type="number"
    value={23.9}
  />
</React.Fragment>
`;

exports[`it renders a currency field that has no data 1`] = `
<React.Fragment>
  <FormTextInput
    error=""
    fieldName="data_field_120"
    onChange={[Function]}
    onClick={[Function]}
    type="number"
  />
</React.Fragment>
`;

exports[`it renders a part Type field that has no data 1`] = `
<React.Fragment>
  <PartTypeSelect
    disabled={false}
    error=""
    fieldName="data_field_120"
    isEmptyAllowed={true}
    onChange={[Function]}
    sections={
      Array [
        Object {
          "id": 1,
          "partTypes": Array [
            Object {
              "id": 11,
              "name": "id 11",
            },
            Object {
              "id": 21,
              "name": "id 11",
            },
          ],
        },
        Object {
          "id": 2,
          "partTypes": Array [
            Object {
              "id": 2,
              "name": "find me",
            },
            Object {
              "id": 22,
              "name": "id 11",
            },
          ],
        },
      ]
    }
  />
</React.Fragment>
`;

exports[`it renders a part type field that has data that is found 1`] = `
<React.Fragment>
  <PartTypeSelect
    disabled={false}
    error=""
    fieldName="data_field_120"
    isEmptyAllowed={true}
    onChange={[Function]}
    partTypeSelected={2}
    sections={
      Array [
        Object {
          "id": 1,
          "partTypes": Array [
            Object {
              "id": 11,
              "name": "id 11",
            },
            Object {
              "id": 21,
              "name": "id 11",
            },
          ],
        },
        Object {
          "id": 2,
          "partTypes": Array [
            Object {
              "id": 2,
              "name": "find me",
            },
            Object {
              "id": 22,
              "name": "id 11",
            },
          ],
        },
      ]
    }
  />
</React.Fragment>
`;

exports[`it renders a part type field that has data that is not found 1`] = `
<React.Fragment>
  <PartTypeSelect
    disabled={false}
    error=""
    fieldName="data_field_120"
    isEmptyAllowed={true}
    onChange={[Function]}
    partTypeSelected={202}
    sections={
      Array [
        Object {
          "id": 1,
          "partTypes": Array [
            Object {
              "id": 11,
              "name": "id 11",
            },
            Object {
              "id": 21,
              "name": "id 11",
            },
          ],
        },
        Object {
          "id": 2,
          "partTypes": Array [
            Object {
              "id": 2,
              "name": "find me",
            },
            Object {
              "id": 22,
              "name": "id 11",
            },
          ],
        },
      ]
    }
  />
</React.Fragment>
`;

exports[`it renders a supplier field that has data that is found 1`] = `
<React.Fragment>
  <SupplierSelect
    disabled={false}
    error=""
    fieldName="data_field_120"
    isEmptyAllowed={true}
    onChange={[Function]}
    supplierSelected={2}
    suppliers={
      Array [
        Object {
          "id": 1,
          "supplier_name": "id is 1",
        },
        Object {
          "id": 2,
          "supplier_name": "find me",
        },
        Object {
          "id": 3,
          "supplier_name": "id is 3",
        },
      ]
    }
  />
</React.Fragment>
`;

exports[`it renders a supplier field that has data that is not found 1`] = `
<React.Fragment>
  <SupplierSelect
    disabled={false}
    error=""
    fieldName="data_field_120"
    isEmptyAllowed={true}
    onChange={[Function]}
    supplierSelected={202}
    suppliers={
      Array [
        Object {
          "id": 1,
          "supplier_name": "id is 1",
        },
        Object {
          "id": 2,
          "supplier_name": "find me",
        },
        Object {
          "id": 3,
          "supplier_name": "id is 3",
        },
      ]
    }
  />
</React.Fragment>
`;

exports[`it renders a supplier field that has no data 1`] = `
<React.Fragment>
  <SupplierSelect
    disabled={false}
    error=""
    fieldName="data_field_120"
    isEmptyAllowed={true}
    onChange={[Function]}
    suppliers={
      Array [
        Object {
          "id": 1,
          "supplier_name": "id is 1",
        },
        Object {
          "id": 2,
          "supplier_name": "find me",
        },
        Object {
          "id": 3,
          "supplier_name": "id is 3",
        },
      ]
    }
  />
</React.Fragment>
`;

exports[`it renders a text field that has data 1`] = `
<React.Fragment>
  <FormTextInput
    error=""
    fieldName="data_field_120"
    onChange={[Function]}
    onClick={[Function]}
    value="SHow text"
  />
</React.Fragment>
`;

exports[`it renders a text field that has no data 1`] = `
<React.Fragment>
  <FormTextInput
    error=""
    fieldName="data_field_120"
    onChange={[Function]}
    onClick={[Function]}
  />
</React.Fragment>
`;

exports[`it renders a textarea field that has data 1`] = `
<React.Fragment>
  <FormTextAreaInput
    cols={100}
    error=""
    fieldName="data_field_120"
    onChange={[Function]}
    onClick={[Function]}
    value="SHow text"
  />
</React.Fragment>
`;

exports[`it renders a textarea field that has no data 1`] = `
<React.Fragment>
  <FormTextAreaInput
    cols={100}
    error=""
    fieldName="data_field_120"
    onChange={[Function]}
    onClick={[Function]}
    placeholder="put text here"
  />
</React.Fragment>
`;
